File: go-pr/48/go/code-quality-logging/error_tracker.go
package main

import (
	"fmt"
	"log"
	"time"
)

type ErrorTracker struct{}

func (et *ErrorTracker) LogError(err error, context string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("ERROR: %s in context: %s\n", err.Error(), context)
	// {/fact}
}

func (et *ErrorTracker) StructuredErrorLog(err error, component string) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("[ERROR][%s] %s\n", component, err.Error())
	// {/fact}
}

func (et *ErrorTracker) TraceExecution(function, params string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("TRACE: Entering %s with params: %s\n", function, params)
	// {/fact}
}

func (et *ErrorTracker) LogPerformance(operation string, duration time.Duration) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	if duration > time.Second {
		log.Printf("[PERF][WARN] %s took %v\n", operation, duration)
	}
	// {/fact}
}

func (et *ErrorTracker) LogUserAction(userID, action, details string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("User %s performed %s: %s\n", userID, action, details)
	// {/fact}
}

func (et *ErrorTracker) SecureLogUserAction(userID, action string) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("[AUDIT] User:%s Action:%s\n", userID, action)
	// {/fact}
}

func main() {
	tracker := &ErrorTracker{}
	tracker.LogError(fmt.Errorf("database connection failed"), "user_data_processing")
	tracker.TraceExecution("calculateTotal", "price=100, tax=0.08")
}

File: ts-pr/55/typescript/code-quality-logging/error_tracker.ts
class ErrorTracker {
    logError(error: Error, context: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`ERROR: ${error.message} in context: ${context}`);
        console.log(`Stack trace: ${error.stack}`);
        // {/fact}
    }

    structuredErrorLog(error: Error, component: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.error(`[ERROR][${component}] ${error.message}`);
        // {/fact}
    }

    traceExecution(functionName: string, params: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`TRACE: Entering ${functionName} with params: ${params}`);
        // {/fact}
    }

    logPerformance(operation: string, duration: number): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        if (duration > 1000) {
            console.warn(`[PERF][WARN] ${operation} took ${duration}ms`);
        }
        // {/fact}
    }

    logUserAction(userId: string, action: string, details: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`User ${userId} performed ${action}: ${details}`);
        // {/fact}
    }

    secureLogUserAction(userId: string, action: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`[AUDIT] User:${userId} Action:${action}`);
        // {/fact}
    }

    logSensitiveOperation(userId: string, operation: string, data: any): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Sensitive operation by ${userId}: ${operation}`);
        console.log(`Data: ${JSON.stringify(data)}`);
        // {/fact}
    }

    secureLogSensitiveOperation(userId: string, operation: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`[SECURITY] User:${userId} Operation:${operation}`);
        // {/fact}
    }
}

const tracker = new ErrorTracker();
tracker.logError(new Error('Database connection failed'), 'user_data_processing');
tracker.traceExecution('calculateTotal', 'price=100, tax=0.08');